### PR TITLE
GRPC adaptations

### PR DESCRIPTION
**Pull request by mykola-kobets-epam**

### CODE REVIEW

File: src/common/crypto/mbedtls/cryptoprovider.cpp [hunk 1]
```diff
@@ -116,6 +117,150 @@ static Error ASN1RemoveTag(const Array<uint8_t>& src, Array<uint8_t>& dst, int t
...
+{
+    result.Resize(result.MaxSize());
+
+    int ret = mbedtls_x509_dn_gets(result.Get(), result.Size(), &dn);
+    if (ret <= 0) {
+        return AOS_ERROR_WRAP(ret);
+    }
+
+    result.Resize(ret);
+    return ErrorEnum::eNone;
```
Comment (mlohvynenko) [inline src/common/crypto/mbedtls/cryptoprovider.cpp:130]: add an empty line before return

File: src/common/crypto/mbedtls/cryptoprovider.cpp [hunk 2]
```diff
@@ -116,6 +117,150 @@ static Error ASN1RemoveTag(const Array<uint8_t>& src, Array<uint8_t>& dst, int t
...
+        return AOS_ERROR_WRAP(ret);
+    }
+
+    mbedtls_entropy_free(&entropy);
+    mbedtls_ctr_drbg_free(&ctrDrbg);
+
+    return ErrorEnum::eNone;
+}
+
+RetWithError<StaticString<16>> ConvertToString(const Time& time)
```
Comment (mlohvynenko) [inline src/common/crypto/mbedtls/cryptoprovider.cpp:160]: should '16' be some kind of constant in header?

File: src/common/crypto/mbedtls/cryptoprovider.cpp [hunk 3]
```diff
@@ -116,6 +117,150 @@ static Error ASN1RemoveTag(const Array<uint8_t>& src, Array<uint8_t>& dst, int t
...
+    // set CA key
+    mbedtls_x509write_crt_set_issuer_key(&clientCert, const_cast<mbedtls_pk_context*>(&caKey));
+
+    // set additional properties: serial, valid time interval
+    ret = mbedtls_x509write_crt_set_serial_raw(&clientCert, const_cast<uint8_t*>(serial.Get()), serial.Size());
+    if (ret != 0) {
+        return AOS_ERROR_WRAP(ret);
+    }
+
+    StaticString<100> notBefore, notAfter;
```
Comment (mlohvynenko) [inline src/common/crypto/mbedtls/cryptoprovider.cpp:230]: should 100 be a constant in header?

File: include/aos/common/tools/string.hpp [hunk 4]
```diff
@@ -473,6 +472,7 @@ class String : public Array<char> {
         return {cDigits[high], cDigits[low]};
     }
 
+private:
```
Comment (al1img) [inline include/aos/common/tools/string.hpp:475]: You have byte array to hex and back. Why do you need this function public? It looks strange, especially with  Pair<char, char>. I can imagine how to use it if it returns char[2] for example. But pair of chars it is really some very specific case.

File: src/common/crypto/mbedtls/cryptoprovider.cpp [hunk 5]
```diff
@@ -269,6 +270,25 @@ Error MbedTLSCryptoProvider::PEMToX509Certs(const String& pemBlob, Array<x509::C
     return ErrorEnum::eNone;
 }
 
+#define PEM_BEGIN_CRT "-----BEGIN CERTIFICATE-----\n"
```
Comment (al1img) [inline src/common/crypto/mbedtls/cryptoprovider.cpp:461]: Do  not use define. `static constexpr auto` as private or probably public member.

Comment (al1img) [inline src/common/crypto/mbedtls/cryptoprovider.cpp:461]: `PEM_BEGIN_CERT`: cert acronym suits better to certificate.

File: src/common/crypto/mbedtls/cryptoprovider.cpp [hunk 6]
```diff
@@ -269,6 +270,25 @@ Error MbedTLSCryptoProvider::PEMToX509Certs(const String& pemBlob, Array<x509::C
...
+#define PEM_END_CRT   "-----END CERTIFICATE-----\n"
+
+Error MbedTLSCryptoProvider::X509CertToPEM(const x509::Certificate& certificate, String& dst)
+{
+    size_t olen;
+
+    auto ret = mbedtls_pem_write_buffer(PEM_BEGIN_CRT, PEM_END_CRT, certificate.mRaw.Get(), certificate.mRaw.Size(),
+        reinterpret_cast<uint8_t*>(dst.Get()), dst.Size(), &olen);
+
```
Comment (al1img) [inline src/common/crypto/mbedtls/cryptoprovider.cpp:470]: Remove empty line

File: src/common/crypto/mbedtls/cryptoprovider.cpp (hunk 2 above)
Comment (al1img) [reply src/common/crypto/mbedtls/cryptoprovider.cpp:160]: It should be public method `ToString` of aos::Time with unit test I guess.

File: tests/utils/src/gencertificates.cmake [hunk 7]
```diff
@@ -0,0 +1,66 @@
+#
```
Comment (al1img) [inline tests/utils/src/gencertificates.cmake:1]: We should move `.cmake` files into `cmake` folder.

File: include/aos/common/crypto/mbedtls/cryptoprovider.hpp [hunk 8]
```diff
@@ -45,6 +45,19 @@ class MbedTLSCryptoProvider : public x509::ProviderItf {
...
+     * Creates certificate chain using client CSR & CA key/certificate as input.
+     *
+     * @param pemCSR client certificate request.
+     * @param pemCAKey CA private key in PEM.
+     * @param pemCACert CA certificate in PEM.
+     * @param serial serial number of certificate.
+     * @param[out] pemClientCert result certificate.
+     * @result Error.
+     */
+    Error CreateClientCert(const String& pemCSR, const String& pemCAKey, const String& pemCACert,
```
Comment (al1img) [inline include/aos/common/crypto/mbedtls/cryptoprovider.hpp:58]: To make it generic this function should have the following API:

```c++
Error CreateClientCert(const x509::CSR& csr, const PrivateKeyItf& caKey, const x509::Certificate& caCert, const Array<uint8_t>& serial, x509::Certificate& clientCert) override;
```

File: include/aos/common/tools/string.hpp (hunk 4 above)
Comment (mykola-kobets-epam) [reply include/aos/common/tools/string.hpp:475]: it is required for percent encoded params in url. 
0x2025u => "%20%25"

File: src/common/crypto/mbedtls/cryptoprovider.cpp (hunk 5 above)
Comment (mykola-kobets-epam) [reply src/common/crypto/mbedtls/cryptoprovider.cpp:461]: I will change. These consts were copy-pasted as is from mbedtls

File: src/common/crypto/mbedtls/cryptoprovider.cpp (hunk 2 above)
Comment (mykola-kobets-epam) [reply src/common/crypto/mbedtls/cryptoprovider.cpp:160]: done

File: src/common/crypto/mbedtls/cryptoprovider.cpp (hunk 5 above)
Comment (mykola-kobets-epam) [reply src/common/crypto/mbedtls/cryptoprovider.cpp:461]: changed

File: src/common/crypto/mbedtls/cryptoprovider.cpp (hunk 6 above)
Comment (mykola-kobets-epam) [reply src/common/crypto/mbedtls/cryptoprovider.cpp:470]: fixed

File: src/common/crypto/mbedtls/cryptoprovider.cpp (hunk 3 above)
Comment (mykola-kobets-epam) [reply src/common/crypto/mbedtls/cryptoprovider.cpp]: fixed

File: include/aos/common/crypto.hpp [hunk 9]
```diff
@@ -483,6 +483,15 @@ class ProviderItf {
      */
     virtual Error PEMToX509Certs(const String& pemBlob, Array<Certificate>& resultCerts) = 0;
 
+    /**
+     * Serializes input Certificate object into a PEM blob.
```
Comment (al1img) [inline include/aos/common/crypto.hpp:501]: `Certificate` -> `certificate`

File: include/aos/common/crypto/mbedtls/cryptoprovider.hpp [hunk 10]
```diff
@@ -54,6 +54,15 @@ class MbedTLSCryptoProvider : public x509::ProviderItf {
      */
     Error PEMToX509Certs(const String& pemBlob, Array<x509::Certificate>& resultCerts) override;
 
+    /**
+     * Serializes input Certificate object into a PEM blob.
```
Comment (al1img) [inline include/aos/common/crypto/mbedtls/cryptoprovider.hpp:71]: ditto

File: include/aos/common/tools/time.hpp [hunk 11]
```diff
@@ -253,6 +259,36 @@ class Time {
         return log;
     }
 
+    /**
+     * Converts time into a string.
+     * Result string has the following format: yyyyMMddhhmmss
+     *
+     * @result RetWithError<StaticString<cTimeStrLength>>.
+     */
+    RetWithError<StaticString<cTimeStrLength>> ToString() const
```
Comment (al1img) [inline include/aos/common/tools/time.hpp:268]: Better to put time conversion into separate commit.

File: src/common/crypto/mbedtls/cryptoprovider.cpp (hunk 5 above)
Comment (al1img) [reply src/common/crypto/mbedtls/cryptoprovider.cpp]: Why don't use them from mbedtls? Are they public in mbedTLS?

File: include/aos/common/tools/time.hpp (hunk 11 above)
Comment (mykola-kobets-epam) [reply include/aos/common/tools/time.hpp:268]: done

File: include/aos/common/config.hpp [hunk 12]
```diff
@@ -471,8 +471,18 @@
...
+ * Maximum number of public keys to be allocated by cryptoprovider simultaneously.
+ */
 #ifndef AOS_CONFIG_CRYPTOPROVIDER_PUB_KEYS_COUNT
 #define AOS_CONFIG_CRYPTOPROVIDER_PUB_KEYS_COUNT 5
 #endif
 
+/**
+ * Size of a time in string representation.
+ */
+#ifndef AOS_TIME_STRING_LENGTH
```
Comment (al1img) [inline include/aos/common/config.hpp:484]: Should be in `common/tools/config.hpp` and named `AOS_CONFIG_TIME_STR_LEN`

File: include/aos/common/crypto/mbedtls/cryptoprovider.hpp (hunk 10 above)
Comment (mykola-kobets-epam) [reply include/aos/common/crypto/mbedtls/cryptoprovider.hpp]: fixed

File: include/aos/common/tools/time.hpp [hunk 13]
```diff
@@ -21,6 +22,11 @@ namespace aos {
  */
 using Duration = int64_t;
 
+/**
+ * Size of a time in string representation.
+ */
+const auto cTimeStrLength = AOS_TIME_STRING_LENGTH;
```
Comment (al1img) [inline include/aos/common/tools/time.hpp:28]: `cTimeStrLen`

File: src/common/crypto/mbedtls/cryptoprovider.cpp [hunk 14]
```diff
@@ -117,6 +117,127 @@ static Error ASN1RemoveTag(const Array<uint8_t>& src, Array<uint8_t>& dst, int t
     return ErrorEnum::eNone;
 }
 
+Error ParseDN(const mbedtls_x509_name& dn, String& result)
```
Comment (MykolaSolyanko) [inline src/common/crypto/mbedtls/cryptoprovider.cpp:120]: this method and below should be static

File: include/aos/common/tools/time.hpp (hunk 13 above)
Comment (mykola-kobets-epam) [reply include/aos/common/tools/time.hpp]: fixed

File: include/aos/common/config.hpp (hunk 12 above)
Comment (mykola-kobets-epam) [reply include/aos/common/config.hpp]: fixed

File: src/common/crypto/mbedtls/cryptoprovider.cpp [hunk 15]
```diff
@@ -228,6 +349,50 @@ Error MbedTLSCryptoProvider::CreateCertificate(
...
+    mbedtls_x509_crt_init(&caCrt);
+    ret = mbedtls_x509_crt_parse(&caCrt, reinterpret_cast<const uint8_t*>(pemCACert.CStr()), pemCACert.Size() + 1);
+    if (ret != 0) {
+        return AOS_ERROR_WRAP(ret);
+    }
+
+    // create client certificate
+    err = aos::crypto::CreateClientCert(csr, caKey, caCrt, serial, pemClientCert);
+    if (!err.IsNone()) {
+        return err;
```
Comment (MykolaSolyanko) [inline src/common/crypto/mbedtls/cryptoprovider.cpp:385]: in this method and method above need to free resource in error case

**Review by al1img — APPROVED** [review]
